File: vw_c/src/Vulkan/Queue.cpp
#include "Vulkan/Queue.h"

#include <span>
#include <type_traits>
#include <VulkanWrapper/Synchronization/Fence.h>
#include <VulkanWrapper/Synchronization/Semaphore.h>
#include <VulkanWrapper/Vulkan/Queue.h>

vw::Fence *vw_queue_submit(vw::Queue *queue,
                           const VwQueueSubmitArguments *arguments) {
    static_assert(std::is_standard_layout_v<vk::CommandBuffer>);
    static_assert(std::is_standard_layout_v<vk::PipelineStageFlagBits>);

    std::span command_buffers = {
        reinterpret_cast<const vk::CommandBuffer *>(arguments->command_buffers),
        arguments->command_buffer_count};

    std::span wait_stages = {reinterpret_cast<const vk::PipelineStageFlags *>(
                                 arguments->wait_stages),
                             arguments->wait_stage_count};

    std::span wait_semaphores = {
        reinterpret_cast<const vk::Semaphore *>(arguments->wait_semaphores),
        arguments->wait_semaphore_count};

    std::span signal_semaphores = {
        reinterpret_cast<const vk::Semaphore *>(arguments->signal_semaphores),
        arguments->signal_semaphores_count};

    return new vw::Fence{
        queue->submit(wait_stages, wait_semaphores, signal_semaphores)};
}

File: vw_rust/src/vulkan/queue.rs
use crate::command::command_buffer::CommandBuffer;
use crate::synchronization::fence::Fence;
use crate::sys::bindings::{
    self, vw_Queue, VkPipelineStageFlagBits, VkSemaphore, VwQueueSubmitArguments,
};
use crate::vulkan::device::Device;

pub struct Queue<'a> {
    ptr: *mut vw_Queue,
    device: &'a Device<'a>,
}

impl<'a> Queue<'a> {
    pub fn new(_device: &'a Device<'a>, ptr: *mut vw_Queue) -> Queue<'a> {
        Queue {
            ptr: ptr,
            device: _device,
        }
    }

    pub fn submit(
        &self,
        cmd_buffers: &[CommandBuffer<'a>],
        wait_stages: &[VkPipelineStageFlagBits],
        wait_semaphores: &[VkSemaphore],
        signal_semaphores: &[VkSemaphore],
    ) -> Fence<'a> {
        let arguments = VwQueueSubmitArguments {
            command_buffers: cmd_buffers.as_ptr() as *const _,
            command_buffer_count: cmd_buffers.len() as u32,
            wait_stages: wait_stages.as_ptr() as *const _,
            wait_stage_count: wait_stages.len() as u32,
            wait_semaphores: wait_semaphores.as_ptr(),
            wait_semaphore_count: wait_semaphores.len() as u32,
            signal_semaphores: signal_semaphores.as_ptr(),
            signal_semaphores_count: signal_semaphores.len() as u32,
        };
        unsafe {
            let fence = bindings::vw_queue_submit(self.ptr, &arguments);
            Fence::new(fence, self.device)
        }
    }
}
